### PR TITLE
[SPARK-8305] [SPARK-8190] [SQL] improve codegen

### PR DESCRIPTION
**Pull request by davies**
This PR fix a few small issues about codgen:
1. cast decimal to boolean
2. do not inline literal with null
3. improve SpecificRow.equals()
4. test expressions with optimized express
5. fix compare with BinaryType

cc @rxin @chenghao-intel

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java [hunk 1]
```diff
@@ -154,6 +154,27 @@ public int fieldIndex(String name) {
...
+  @Override
+  public boolean equals(Object other) {
+    if (other instanceof Row) {
+      Row row = (Row) other;
+      int n = size();
+      if (n != row.size()) {
+        return false;
+      }
+      for (int i = 0; i < n; i ++) {
+        if (isNullAt(i) != row.isNullAt(i) || !isNullAt(i) && !get(i).equals(row.get(i))) {
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java]: Will that cause NPE for `!get(i).equals(row.get(i))`

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java]: Sorry, ignore the above comment. But it's better to add a bracket for `!isNullAt(i) && !get(i).equals(row.get(i))`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -161,15 +161,22 @@ class CodeGenContext {
...
+    case dt: DataType if isPrimitiveType(dt) => s"$c1 == $c2"
+    case other => s"$c1.equals($c2)"
+  }
+
+  /**
+   * Generate code for compare expression in Java
+   */
+  def genComp(dataType: DataType, c1: String, c2: String): String = dataType match {
+    case FloatType | DoubleType => s"(int)java.lang.Math.signum($c1 - $c2)"
+    case dt: DataType if isPrimitiveType(dt) => s"(int)($c1 - $c2)"
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: `(1.2 - 1.1).asInstanceOf[Int]` => 0

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: The first case will handle float and double, using java.lang.Math.signum

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: can you add inline comment explaining why we have special case for float/double?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java (hunk 1 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java]: +1 on adding parentheses.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 3]
```diff
@@ -250,15 +250,14 @@ abstract class BinaryComparison extends BinaryExpression with Predicate {
   }
 
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
-    left.dataType match {
-      case dt: NumericType if ctx.isNativeType(dt) => defineCodeGen (ctx, ev, {
-        (c1, c3) => s"$c1 $symbol $c3"
+    if (ctx.isPrimitiveType(left.dataType)) {
+      // faster version
+      defineCodeGen(ctx, ev, {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: can we just do
```
defineCodeGen(ctx, ev, (c1, c2) => s"$c1 $symbol $c2")
```
in one line for everything here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 4]
```diff
@@ -281,7 +280,7 @@ case class EqualTo(left: Expression, right: Expression) extends BinaryComparison
     else java.util.Arrays.equals(l.asInstanceOf[Array[Byte]], r.asInstanceOf[Array[Byte]])
   }
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: add a space here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala [hunk 5]
```diff
@@ -53,4 +53,13 @@ object TypeUtils {
 
   def getOrdering(t: DataType): Ordering[Any] =
     t.asInstanceOf[AtomicType].ordering.asInstanceOf[Ordering[Any]]
+
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala]: remove one extra line

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala [hunk 6]
```diff
@@ -53,4 +53,13 @@ object TypeUtils {
...
+
+  def compareBinary(x: Array[Byte], y: Array[Byte]): Int = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala]: this seems super slow ... after this PR we should create a ByteArrayUtils in unsafe.types

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala (hunk 5 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala]: done

**Pull request closed without merging**